Source/Plugins/Core/com.equella.core/js/tsrc/settings/SchemaSelector.tsx

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * The Apereo Foundation licenses this file to you under the Apache License,
+ * Version 2.0, (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as React from "react";
+import { ReactElement } from "react";
+import { Grid, InputLabel, MenuItem, Select } from "@material-ui/core";
+import {
+  schemaListSummary,
+  SchemaNode,
+  schemaTree,
+} from "../schema/SchemaModule";
+import SchemaNodeSelector from "../settings/SchemaNodeSelector";
+import { languageStrings } from "../util/langstrings";
+
+interface SchemaSelectorProps {
+  /**
+   * Callback function triggered upon selection of a node.
+   * @param node  Path of the selected node.
+   */
+  setSchemaNode: (node: string) => void;
+}
+
+/**
+ * This component defines a schema selector, for selecting a schema and then a node within.
+ * When a schema is selected, it will display that schema within a SchemaNodeSelector.
+ *
+ */
+export default function SchemaSelector({ setSchemaNode }: SchemaSelectorProps) {
+  const [selectedSchema, setSelectedSchema] = React.useState<
+    string | undefined
+  >(undefined);
+  const [schema, setSchema] = React.useState<SchemaNode>();
+  const [schemaList, setSchemaList] = React.useState<ReactElement[]>([]);
+  const [schemaNodePath, setSchemaNodePath] = React.useState<string>("");
+  const strings =
+    languageStrings.settings.searching.facetedsearchsetting.schemaselector;
+  React.useEffect(() => {
+    schemaListSummary().then((schemas) => {
+      const elementList: JSX.Element[] = [
+        <MenuItem id={strings.selectaschema} value={undefined}>
+          {strings.selectaschema}
+        </MenuItem>,
+      ];
+      schemas.forEach((name, uuid) => {
+        elementList.push(
+          <MenuItem id={uuid} value={uuid}>
+            {name}
+          </MenuItem>
+        );
+      });
+      setSchemaList(elementList);
```

Comment:
use `map` as an alternative `forEach` with `push` when possible.
`map` makes it visually easier to track data flow

```suggestion
      const defaultListItem = [
        <MenuItem id={strings.selectaschema} value={undefined}>
          {strings.selectaschema}
        </MenuItem>,
      ];
      const schemaListItems = schemas.map((name, uuid) => (
        <MenuItem id={uuid} value={uuid}>
          {name}
        </MenuItem>
      ));
      setSchemaList(defaultListItem.concat(schemaListItems));
```

Reply:
And to just flip it around slightly. If you can simply add to your `schemas` with the default value at the top, then you could simply have:

```typescript
setSchemaList(schemas.map((name, uuid) => (<MenuItem>...</MenuItem>));
```

Reply:
.map is not a function of Map<string, string> in Typescript. At least, my IDE doesn't think so. Neither of these work.